packages/appium/docs/en/ecosystem/build-drivers.md

```diff
@@ -2,4 +2,617 @@
+WebDriver Protocol, plus the various Appium extensions, has a corresponding function that you
+implement if you want to support that command in your driver. The best way to see which commands
+Appium supports and which method you need to implement for each command is to look at Appium's
+[routes.js](https://github.com/appium/appium/blob/2.0/packages/base-driver/lib/protocol/routes.js).
+Each route object in this file tells you the command name as well as the parameters you'd expect to
+receive for that command.
+
+Let's take this block for example:
+```js
+'/session/:sessionId/url': {
+    GET: {command: 'getUrl'},
+    POST: {command: 'setUrl', payloadParams: {required: ['url']}},
+}
+```
+
+Here we see that the route `/session/:sessionId/url` is mapped to two commands, one for a `GET`
+request and one for a `POST` request. If we want to allow our driver to change the "url" (or
+whatever that might mean for our driver), we can therefore implement the `setUrl` command, knowing
+it will take the `url` parameter:
+
+```
+async setUrl(url) {
+    // your implementation here
+}
+```
+
+A few notes:
+- all commands should be `async` functions
+- you don't need to worry about protocol encoding/decoding. You will get JS objects as params, and
+  can return JS objects in response. Appium will take care of wrapping it up in the WebDriver
+  protocol response format, turning it into JSON, etc...
+- all session-based commands receive the `sessionId` parameter as the last parameter
+- all element-based commands receive the `elementId` parameter as the second-to-last parameter
+- if your driver doesn't implement a command, users can still try to access the command, and will
+  get a `501 Not Yet Implemented` response error.
+
+### Implement element finding
+
+Element finding is a special command implementation case. You don't actually want to override
+`findElement` or `findElements`, even though those are what are listed in `routes.js`. Appium does
+a lot of work for you if instead you implement this function:
+
+```
+async findElOrEls(strategy, selector, mult, context) {
+    // find your element here
+}
+```
+
+Here's what gets passed in:
+
+- `strategy` - a string, the locator strategy being used
+- `selector` - a string, the selector
+- `mult` - boolean, whether the user has requested one element or all elements matching the
+  selector
+- `context` - (optional) if defined, will be a W3C Element (i.e., a JS object with the W3C element
+  identifier as the key and the element ID as the value)
+
+And you need to return one of the following:
+
+- a single W3C element (an object as described above)
+- an array of W3C elements
+
+Note that you can import that W3C web element identifier from `appium/support`:
+
+```js
+import {util} from 'appium/support';
+const { W3C_WEB_ELEMENT_IDENTIFIER } = util;
+```
+
+What you do with elements is up to you! Usually you end up keeping a cache map of IDs to actual
```

Comment:
should we also mention the set of standard errors we have to map w3c error codes?

Reply:
Sounds good idea

Reply:
yeah, will add